drivers/metal/rendering_device_driver_metal.mm

```diff
@@ -0,0 +1,3919 @@
+		bin_data.spirv_specialization_constants_ids_mask |= (1 << spirv_sc.constant_id);
+	}
+
+	// Reflection using SPIRV-Cross:
+	// https://github.com/KhronosGroup/SPIRV-Cross/wiki/Reflection-API-user-guide
+
+	CompilerMSL::Options msl_options{};
+	msl_options.set_msl_version(version_major, version_minor);
+	if (version_major == 3 && version_minor >= 1) {
+		// TODO(sgc): restrict to Metal 3.0 for now, until bugs in SPIRV-cross image atomics are resolved
+		msl_options.set_msl_version(3, 0);
+	}
+	bin_data.msl_version = msl_options.msl_version;
+#if TARGET_OS_OSX
+	msl_options.platform = CompilerMSL::Options::macOS;
+#else
+	msl_options.platform = CompilerMSL::Options::iOS;
+#endif
+
+#if TARGET_OS_IOS
+	msl_options.ios_use_simdgroup_functions = (*metal_device_properties).features.simdPermute;
+#endif
+
+	msl_options.argument_buffers = true;
+	msl_options.force_active_argument_buffer_resources = true; // same as MoltenVK when using argument buffers
+	// msl_options.pad_argument_buffer_resources = true; // same as MoltenVK when using argument buffers
+	msl_options.texture_buffer_native = true; // texture_buffer support
+	msl_options.use_framebuffer_fetch_subpasses = false;
+	msl_options.pad_fragment_output_components = true;
+	msl_options.r32ui_alignment_constant_id = R32UI_ALIGNMENT_CONSTANT_ID;
+	msl_options.agx_manual_cube_grad_fixup = true;
+
+	CompilerGLSL::Options options{};
+	options.vertex.flip_vert_y = true;
+#if DEV_ENABLED
+	options.emit_line_directives = true;
+#endif
+
+	for (uint32_t i = 0; i < p_spirv.size(); i++) {
+		ShaderStageSPIRVData const &v = p_spirv[i];
+		ShaderStage stage = v.shader_stage;
+		char const *stage_name = SHADER_STAGE_NAMES[stage];
+		uint32_t const *const ir = reinterpret_cast<uint32_t const *const>(v.spirv.ptr());
+		size_t word_count = v.spirv.size() / sizeof(uint32_t);
+		Parser parser(ir, word_count);
+		try {
+			parser.parse();
+		} catch (CompilerError &e) {
+			ERR_FAIL_V_MSG(Result(), "Failed to parse IR at stage " + String(SHADER_STAGE_NAMES[stage]) + ": " + e.what());
+		}
+
+		CompilerMSL compiler(std::move(parser.get_parsed_ir()));
+		compiler.set_msl_options(msl_options);
+		compiler.set_common_options(options);
+
+		std::unordered_set<VariableID> active = compiler.get_active_interface_variables();
+		ShaderResources resources = compiler.get_shader_resources();
+
+		std::string source = compiler.compile();
+
+		ERR_FAIL_COND_V_MSG(compiler.get_entry_points_and_stages().size() != 1, Result(), "Expected a single entry point and stage.");
+
+		EntryPoint &entry_point_stage = compiler.get_entry_points_and_stages().front();
+		SPIREntryPoint &entry_point = compiler.get_entry_point(entry_point_stage.name, entry_point_stage.execution_model);
+
+		// process specialization constants
+		if (!compiler.get_specialization_constants().empty()) {
+			for (SpecializationConstant const &constant : compiler.get_specialization_constants()) {
+				LocalVector<SpecializationConstantData>::Iterator res = std::find_if(
```

Comment:
Just spotted this. I believe that, given STL is mostly fordidden in Godot, we'd need the counterparts of some useful algos like that. In the meantime, you have to use old-fashioned code for that.

Reply:
@RandomShaper I've removed `std::find_if` and `std::for_each` calls.